doc/pcre2_compile.3

```diff
@@ -65,6 +65,7 @@ The primary option bits are:
                             theses (named ones available)
   PCRE2_NO_AUTO_POSSESS    Disable auto-possessification
   PCRE2_NO_DOTSTAR_ANCHOR  Disable automatic anchoring for .*
+  PCRE2_NO_PATTERN_REWRITE Disable pattern rewriting optimizations
```

Comment:
In the future, there might be more rewrite optimizations. I would name this specifically for https://github.com/PCRE2Project/pcre2/issues/411 optimization only.

Reply:
Disagree (at least in principle). As described in the discussion for that feature request this is indeed adding a completely new mode of operation for PCRE which was NEVER an optimizing compiler and where the normal use cases expected some other way to optimize the pattern (the same applies to JIT BTW which mostly translates the interpreter opcodes into native CPU code)

I "agree" that long term other "rewrites" might be added and would be a good idea to identify them here, but the way to do so would be to add flags for each case as extended options IMHO.  Indeed I would argue that this option SHOULDN'T be using one of the few options bits left and should be instead enabled through an extended option itself.

Reply:
I would use an O0-O3 system similar to gcc. The O1 is the default, and the behaviour is same as now. Then we could introduce optimizations, which costs more time, but improves performace more. We could also disable optimizations, such as auto-possession with O0. Disabling first character optimizations would be deprecated as well.

Reply:
My view when Alex and I discussed this was that NO_PATTERN_REWRITE would disable *all* rewriting optimizations, present and future. If there is a need to disable individual ones, then additional extended options can be used.

Reply:
I think I wasn't clear enough and indeed might had confused the discussion further by mentioning the extended options, which I was proposing should be used to ENABLE each optimization, and therefore the `#define` and bit in the compile options that this thread is attached two could be reused.

Agree that the `(*NO_PATTERN_REWRITE)`  pattern option setting should disable all optimizations.

Reply:
What do you think about `(*OPT0)` - `(*OPT3)`? People in general don't really understand the effect of individual optimizations. O0-O3 is a simple way to say I don't mind a longer compilation time for faster execution.

We could add a new api to set the optimization level in compile context.

Reply:
I guess that setting an optimization level in pattern with `(*OPT=n)` might be useful to end users that have a new enough library and no other way to enable them, but still think it would be better to enable them through an API call instead (or an extended option) based on the use cases spelled in the original request.

I also think it will be better to be disabled by default (your idea of level 1 being the default and including "autoposification" as a "rewrite like" optimization if you squint hard enough, might work) but I suspect it might be cleared to start with 0 and had this part of 1.

Reply:
Could you describe your suggestion for the whole API? Including C and pattern string API. I think this is the most important part to decide, the rest can be changed internally between releases.

Reply:
My hope was that this new optimization would be cheap enough that it could happen by default, just like pcre2_study(). There are two reasons for having a disabling option: (1) As an emergency safety net in case there are broken cases and (2) so that it's easy to make comparisons with and without the optimization. I am not really keen on optimization levels.

Reply:
We had yet to see any performance numbers though, and I am also a little concerned by the increase on code size memory from the test output, but it might be just bugs and easy to nail down once we are back to having code that builds mostly warning free and without leaks.

Going back to Zoltan's API question, are you suggesting we punt the need for an API or extra flags until there is another optimization? or even until there are problems with one of them specifically?

Reply:
> We [have] yet to see any performance numbers though...

That will be addressed soon enough. I am dealing with the issues raised one by one.

Reply:
I would like all current and future optimization handled in the same way. There are only a few compile flags remained, and I would prefer reserving them for features, rather than optimizations. If we add 2-3 optimizations / year, that will be 20-30 in 10 years. Would be good if their generic interface is decided now, rather than each optimization is handled by different ways.

Reply:
> My hope was that this new optimization would be cheap enough that it could happen by default, just like pcre2_study()

The problem though is that unlike `pcre_study()` the pattern itself could change and therefore basic things like `erroroffset` become unreliable and as pointed out in #460 could be even a source of crashes (another reason why I think it might not be a good idea to allow it by default).

Eitherway I agree with Zoltan that nailing the API first is worth doing with the first release (even better if with the first commit)

Reply:
> The problem though is that unlike the `pcre*_study()` optimizations, the pattern itself could change and therefore basic things like `erroroffset` become unreliable and as pointed out in #460 could be even a source of crashes (another reason why I think it might not be a good idea to allow it by default).

What you have mentioned here is a very serious issue. I may put this PR 'on hold' to allow time to think about this. It could be a fatal flaw.

Reply:
With regard to the `erroroffset` issue raised by @carenas: I think that with the existing PCRE2 API, pattern rewriting optimizations of the kind in this PR can be done *only* if all possible causes of errors (or at least, errors which require a value to be returned in `erroroffset`) are checked first. Pattern rewriting can only be done once we are sure that there is no possible reason to return an `erroroffset`.

Reply:
Ah... 🤦🏻‍♂️ Sorry, scratch the above comment.

I've just remembered: the pattern rewriting phase added in this PR does not rewrite the *pattern string*; it rewrites the `parsed_pattern` vector which comes from the pattern string.

Some of the `parsed_pattern` items have associated data which includes offsets into the original pattern string; when adjusting the `parsed_pattern` vector, the pattern rewriter maintains all these offsets "as is", so that if any error is detected later in the compilation process, `erroroffset` will point to the appropriate part of the original pattern string.

(This might be a good debug assertion: run through the `parsed_pattern` vector and check that all pattern string offsets are within bounds.)

Reply:
I thought the rewriting happens after the parsing is successful, so no error (besides out of memory) can happen.

Reply:
> I thought the rewriting happens after the parsing is successful, so no error (besides out of memory) can happen.

You are right that rewriting happens after a successful parse... but there are just a few types of errors which are not detected during parsing, but rather when processing the `parsed_pattern` vector. For all such cases, offsets into the original pattern string are stored in the `parsed_pattern` vector, and the pattern rewriter makes sure not to destroy these.

Reply:
The erroroffset issue was likely a red herring, but it is offtopic for this conversation which I think needs a resolution.

It would seem that @PhilipHazel likes the current one as it is sufficient and doesn't make a difference when there is only one rewriting optimization anyway and might want this discussion punted until later.

Note that the option flag mentioned here is to disable all rewriting.

Reply:
Whatever we do, it should happen before the next release. There are only a few compile flags left, and this flag feel out of scope for me. What about making this always enabled in this patch, and make it conditional when we decide how to do it.

Reply:
Definitely, this issue needs to be resolved. But even just for testing purposes, it's vital that we can disable the optimization. I think making it "always enabled" is not an option.

My next priority on this PR is to investigate why the 'code memory' size has increased for some patterns. After that will be to eliminate all compiler warnings. Other issues to be worked on include: enabling tests on Windows if appropriate, coming to a resolution on whether PCRE2 will move to the C99 standard or not, questions of brace style, and performance benchmarking. (Philip has already done a bit of benchmarking, and from his limited results, it appeared that this optimization is both 1) worthwhile and 2) has relatively small impact on compile time. But more thorough and extensive benchmarking is needed. It's possible that the benchmarking might cause me to do more performance work to reduce the impact on compile time.)

As all those matters fall into place one by one, at some point, the issue of how to disable the optimization must be worked out. I agree with Zoltan that the number of bits left in the compile options is truly small. But otherwise I haven't thought hard enough to weigh in on the issue. I just followed what has been done for previous optimizations.

I consider it extremely important that Zoltan, Carlo, and other key contributors who have helped to make PCRE2 what it is now are all happy with whatever is decided.